.circleci/config.yml

```diff
@@ -0,0 +1,438 @@
+
+  environment: &envvars
+    # required by some of the targets, e.g. compcert, passed for
+    # instance to opam to configure the number of parallel jobs
+    # allowed
+    NJOBS: 2
+    COMPILER: "system"
+    CAMLP5_VER: "6.14"
+    NATIVE_COMP: "yes"
+
+    # some useful values
+    COMPILER_32BIT: &compiler-32bit "4.02.3+32bit"
+
+    COMPILER_BLEEDING_EDGE: &compiler-be "4.06.0"
+    CAMLP5_VER_BLEEDING_EDGE: &camlp5-ver-be "7.03"
+
+    TIMING_PACKAGES: &timing-packages "time python"
+
+    COQIDE_PACKAGES: &coqide-packages "libgtk2.0-dev libgtksourceview2.0-dev"
+    #COQIDE_PACKAGES_32BIT: "libgtk2.0-dev:i386 libgtksourceview2.0-dev:i386"
+    COQIDE_OPAM: &coqide-opam "lablgtk-extras"
+    COQIDE_OPAM_BE: &coqide-opam-be "num lablgtk.2.18.6 lablgtk-extras.1.6"
+    COQDOC_PACKAGES: &coqdoc-packages "texlive-latex-base texlive-latex-recommended texlive-latex-extra texlive-math-extra texlive-fonts-recommended texlive-fonts-extra latex-xcolor ghostscript transfig imagemagick tipa"
+    COQDOC_OPAM: &coqdoc-opam "hevea"
+
+version: 2
+
+before_script: &before_script
+  name: before_script
+  command: |
+    source ~/.profile
+    export TERM=xterm
+    ls # figure out if artifacts are around
+    printenv
+    #if [ "$COMPILER" = "$COMPILER_32BIT" ]; then sudo dpkg --add-architecture i386; fi
+    if [ -n "${EXTRA_PACKAGES}" ]; then sudo apt-get update -yq && sudo apt-get install -yq --no-install-recommends ${EXTRA_PACKAGES}; fi
+
+    # the default repo in this docker image is a local directory
+    # at the time of 4aaeb8abf it lagged behind the official
+    # repository such that camlp5 7.01 was not available
+    opam repository set-url default https://opam.ocaml.org
+    opam update
+    opam switch ${COMPILER}
+    eval $(opam config env)
+    opam config list
+    opam install -j ${NJOBS} -y camlp5.${CAMLP5_VER} ocamlfind ${EXTRA_OPAM}
+    rm -rf ~/.opam/log/
+    opam list
+
+.build-template: &build-template
+  <<: *params
+  steps:
+    - checkout
+    - restore_cache:
+        keys:
+          - coq-opam-cache-v1-{{ arch }}-{{ checksum ".circleci/config.yml" }}-
+          - coq-opam-cache-v1-{{ arch }}- # this grabs old cache if checksum doesn't match
+    - run: *before_script
+    - save_cache:
+        key: coq-opam-cache-v1-{{ arch }}-{{ checksum ".circleci/config.yml" }}-
+        paths:
+          - ~/.opam
+    - run: &build-configure
+        name: Configure
+        command: |
+          source ~/.profile
+
+          echo 'start:coq.config'
+          ./configure -local -native-compiler ${NATIVE_COMP} ${EXTRA_CONF}
+          echo 'end:coq.config'
+    - run: &build-build
+        name: Build
+        command: |
+          source ~/.profile
+          echo 'start:coq.build'
+          make -j ${NJOBS} byte
+          make -j ${NJOBS}
+          make test-suite/misc/universes/all_stdlib.v
+          echo 'end:coq:build'
+    - persist_to_workspace:
+        root: &workspace ~/
+        paths:
+          - coq/
```

Comment:
I think the `.opam` should be in the workspace. This will avoid the dance with the cache in the other targets.